In [1]:
from carbon import CarbonSimulatorUI, __version__, __date__
print(f"Carbon Version v{__version__} ({__date__})", )

Carbon Version v2.0-beta2 (8/Dec/2022)


# Carbon Simulation - Demo 4-2

In this demo we investigate the **Alpha Router performance** we add dust lmiit orders at either end of the main strats

Initialize an exact simulator and an alpha simulator

In [2]:
Sim = CarbonSimulatorUI(pair="ETH/USDC", verbose=False)
AlphaSim = CarbonSimulatorUI(pair="ETH/USDC", verbose=False, matching_method=CarbonSimulatorUI.MATCH_ALPHA)

Add identical strategies to both simulators

In [3]:
strat = ['ETH', 10, 2000, 2500, 1000, 2800, 2700]
dusta = ['ETH', 0.00001, 2050, 2050, 0.00001, 2850, 2850]
dustb = ['ETH', 0.00001, 2550, 2550, 0.00001, 2750, 2750]
dustc = ['ETH', 0.00001, 1950, 1950, 0.00001, 2750, 2750]
dustd = ['ETH', 0.00001, 2650, 2650, 0.00001, 2650, 2650]
for _ in range(15):
    Sim.add_strategy(*strat)
    AlphaSim.add_strategy(*strat)
    Sim.add_strategy(*dusta)
    AlphaSim.add_strategy(*dusta)
    Sim.add_strategy(*dustb)
    AlphaSim.add_strategy(*dustb)
    Sim.add_strategy(*dustc)
    AlphaSim.add_strategy(*dustc)
    Sim.add_strategy(*dustd)
    AlphaSim.add_strategy(*dustd)

#Sim.state()['orders']
AlphaSim.state()['orders']

id     pair   tkn       y_int           y y_unit  disabled  p_start  \
0      0  ETHUSDC   ETH    10.00000    10.00000    ETH     False   2000.0   
1      1  ETHUSDC  USDC  1000.00000  1000.00000   USDC     False   2800.0   
2      2  ETHUSDC   ETH     0.00001     0.00001    ETH     False   2050.0   
3      3  ETHUSDC  USDC     0.00001     0.00001   USDC     False   2850.0   
4      4  ETHUSDC   ETH     0.00001     0.00001    ETH     False   2550.0   
..   ...      ...   ...         ...         ...    ...       ...      ...   
145  145  ETHUSDC  USDC     0.00001     0.00001   USDC     False   2750.0   
146  146  ETHUSDC   ETH     0.00001     0.00001    ETH     False   1950.0   
147  147  ETHUSDC  USDC     0.00001     0.00001   USDC     False   2750.0   
148  148  ETHUSDC   ETH     0.00001     0.00001    ETH     False   2650.0   
149  149  ETHUSDC  USDC     0.00001     0.00001   USDC     False   2650.0   

      p_end  p_marg        p_unit  lid  
0    2500.0  2000.0  USDC per ETH    1  
1    2700.0  2800.0  USDC per ETH    0  
2    2050.0  2050.0  USDC per ETH    3  
3    2850.0  2850.0  USDC per ETH    2  
4    2550.0  2550.0  USDC per ETH    5  
..      ...     ...           ...  ...  
145  2750.0  2750.0  USDC per ETH  144  
146  1950.0  1950.0  USDC per ETH  147  
147  2750.0  2750.0  USDC per ETH  146  
148  2650.0  2650.0  USDC per ETH  149  
149  2650.0  2650.0  USDC per ETH  148  

[150 rows x 12 columns]

In [4]:
AlphaSim.state()['orders'].query('id==18')

id     pair  tkn    y_int        y y_unit  disabled  p_start   p_end  \
18  18  ETHUSDC  ETH  0.00001  0.00001    ETH     False   2650.0  2650.0   

    p_marg        p_unit  lid  
18  2650.0  USDC per ETH   19

In [5]:
AlphaSim.state()['orders'].query('id==36')

id     pair  tkn    y_int        y y_unit  disabled  p_start   p_end  \
36  36  ETHUSDC  ETH  0.00001  0.00001    ETH     False   1950.0  1950.0   

    p_marg        p_unit  lid  
36  1950.0  USDC per ETH   37

In [6]:
# set test amounts
usdc_trade_amount = 1000
eth_trade_amount = 0.5

## Route by Source

#### AMM buys USDC

First we test the standard simulator which should route through all 5 positions

In [7]:
Sim.amm_buys('USDC',usdc_trade_amount, execute=False)['trades']  # route_trade_by_source

uid  id subid                          note   aggr   exec limitfail  \
0   0.0   0     0                      route #6  False  False      None   
0   0.1   0     1                     route #16  False  False      None   
0   0.2   0     2                     route #26  False  False      None   
0   0.3   0     3                     route #36  False  False      None   
0   0.4   0     4                     route #46  False  False      None   
0   0.5   0     5                     route #56  False  False      None   
0   0.6   0     6                     route #66  False  False      None   
0   0.7   0     7                     route #76  False  False      None   
0   0.8   0     8                     route #86  False  False      None   
0   0.9   0     9                     route #96  False  False      None   
0  0.10   0    10                    route #106  False  False      None   
0  0.11   0    11                    route #116  False  False      None   
0  0.12   0    12                    route #126  False  False      None   
0  0.13   0    13                    route #136  False  False      None   
0  0.14   0    14                    route #146  False  False      None   
0  0.15   0    15                      route #0  False  False      None   
0  0.16   0    16                     route #10  False  False      None   
0  0.17   0    17                     route #20  False  False      None   
0  0.18   0    18                     route #30  False  False      None   
0  0.19   0    19                     route #40  False  False      None   
0  0.20   0    20                     route #50  False  False      None   
0  0.21   0    21                     route #60  False  False      None   
0  0.22   0    22                     route #70  False  False      None   
0  0.23   0    23                     route #80  False  False      None   
0  0.24   0    24                     route #90  False  False      None   
0  0.25   0    25                    route #100  False  False      None   
0  0.26   0    26                    route #110  False  False      None   
0  0.27   0    27                    route #120  False  False      None   
0  0.28   0    28                    route #130  False  False      None   
0  0.29   0    29                    route #140  False  False      None   
0     0   0     A  AMM sells 0ETH buys 1000USDC   True  False      None   

       amt1 tkn1         amt2  tkn2     pair  \
0  0.000010  ETH     0.019500  USDC  ETHUSDC   
0  0.000010  ETH     0.019500  USDC  ETHUSDC   
0  0.000010  ETH     0.019500  USDC  ETHUSDC   
0  0.000010  ETH     0.019500  USDC  ETHUSDC   
0  0.000010  ETH     0.019500  USDC  ETHUSDC   
0  0.000010  ETH     0.019500  USDC  ETHUSDC   
0  0.000010  ETH     0.019500  USDC  ETHUSDC   
0  0.000010  ETH     0.019500  USDC  ETHUSDC   
0  0.000010  ETH     0.019500  USDC  ETHUSDC   
0  0.000010  ETH     0.019500  USDC  ETHUSDC   
0  0.000010  ETH     0.019500  USDC  ETHUSDC   
0  0.000010  ETH     0.019500  USDC  ETHUSDC   
0  0.000010  ETH     0.019500  USDC  ETHUSDC   
0  0.000010  ETH     0.019500  USDC  ETHUSDC   
0  0.000010  ETH     0.019500  USDC  ETHUSDC   
0  0.033312  ETH    66.647167  USDC  ETHUSDC   
0  0.033312  ETH    66.647167  USDC  ETHUSDC   
0  0.033312  ETH    66.647167  USDC  ETHUSDC   
0  0.033312  ETH    66.647167  USDC  ETHUSDC   
0  0.033312  ETH    66.647167  USDC  ETHUSDC   
0  0.033312  ETH    66.647167  USDC  ETHUSDC   
0  0.033312  ETH    66.647167  USDC  ETHUSDC   
0  0.033312  ETH    66.647167  USDC  ETHUSDC   
0  0.033312  ETH    66.647167  USDC  ETHUSDC   
0  0.033312  ETH    66.647167  USDC  ETHUSDC   
0  0.033312  ETH    66.647167  USDC  ETHUSDC   
0  0.033312  ETH    66.647167  USDC  ETHUSDC   
0  0.033312  ETH    66.647167  USDC  ETHUSDC   
0  0.033312  ETH    66.647167  USDC  ETHUSDC   
0  0.033312  ETH    66.647167  USDC  ETHUSDC   
0  0.499828  ETH  1000.000000  USDC  ETHUSDC   

                                             routeix  nroutes     

Then we can set a threshold on the number of orders using the AlphaSim

In [8]:
AlphaSim.amm_buys('USDC',usdc_trade_amount, execute=False, threshold_orders=3)['trades']  # route_trade_by_source

uid  id subid                          note   aggr   exec limitfail  \
0  0.0   0     0                     route #16  False  False      None   
0  0.1   0     1                     route #26  False  False      None   
0  0.2   0     2                      route #0  False  False      None   
0    0   0     A  AMM sells 0ETH buys 1000USDC   True  False      None   

       amt1 tkn1       amt2  tkn2     pair      routeix  nroutes        price  \
0  0.000010  ETH     0.0195  USDC  ETHUSDC           16        1  2010.554434   
0  0.000010  ETH     0.0195  USDC  ETHUSDC           26        1  2010.554434   
0  0.497355  ETH   999.9610  USDC  ETHUSDC            0        1  2010.554434   
0  0.497375  ETH  1000.0000  USDC  ETHUSDC  [16, 26, 0]        3  2010.554434   

         p_unit  threshold_orders  
0  USDC per ETH                 3  
0  USDC per ETH                 3  
0  USDC per ETH                 3  
0  USDC per ETH                 3

We would expect that routing through fewer orders should give a slightly worse price for the trader.

Since here the amm is buying then the trader is selling 1000 USDC, then the Alpha buy amount of 0.499122 is slightly lower than the 0.49973 ETH achieved in exact routing. However, the difference here is only 7bp

In [9]:
AlphaSim.state()["trades"].query("uid=='0'").iloc[0]["amt1"],Sim.state()["trades"].query("uid=='0'").iloc[0]["amt1"]

(0.497375, 0.499828)

In [10]:
10000*(AlphaSim.state()["trades"].query("uid=='0'").iloc[0]["amt1"]
       /Sim.state()["trades"].query("uid=='0'").iloc[0]["amt1"]-1)

-49.07688244756181

#### AMM buys ETH

In [11]:
Sim.amm_buys('ETH',eth_trade_amount, execute=False)['trades']  # route_trade_by_source

uid  id subid                          note   aggr   exec limitfail  \
0   1.0   1     0                      route #3  False  False      None   
0   1.1   1     1                     route #13  False  False      None   
0   1.2   1     2                     route #23  False  False      None   
0   1.3   1     3                     route #33  False  False      None   
0   1.4   1     4                     route #43  False  False      None   
0   1.5   1     5                     route #53  False  False      None   
0   1.6   1     6                     route #63  False  False      None   
0   1.7   1     7                     route #73  False  False      None   
0   1.8   1     8                     route #83  False  False      None   
0   1.9   1     9                     route #93  False  False      None   
0  1.10   1    10                    route #103  False  False      None   
0  1.11   1    11                    route #113  False  False      None   
0  1.12   1    12                    route #123  False  False      None   
0  1.13   1    13                    route #133  False  False      None   
0  1.14   1    14                    route #143  False  False      None   
0  1.15   1    15                      route #1  False  False      None   
0  1.16   1    16                     route #11  False  False      None   
0  1.17   1    17                     route #21  False  False      None   
0  1.18   1    18                     route #31  False  False      None   
0  1.19   1    19                     route #41  False  False      None   
0  1.20   1    20                     route #51  False  False      None   
0  1.21   1    21                     route #61  False  False      None   
0  1.22   1    22                     route #71  False  False      None   
0  1.23   1    23                     route #81  False  False      None   
0  1.24   1    24                     route #91  False  False      None   
0  1.25   1    25                    route #101  False  False      None   
0  1.26   1    26                    route #111  False  False      None   
0  1.27   1    27                    route #121  False  False      None   
0  1.28   1    28                    route #131  False  False      None   
0  1.29   1    29                    route #141  False  False      None   
0     1   1     A  AMM sells 1398USDC buys 0ETH   True  False      None   

          amt1  tkn1      amt2 tkn2     pair  \
0     0.000010  USDC  0.000000  ETH  ETHUSDC   
0     0.000010  USDC  0.000000  ETH  ETHUSDC   
0     0.000010  USDC  0.000000  ETH  ETHUSDC   
0     0.000010  USDC  0.000000  ETH  ETHUSDC   
0     0.000010  USDC  0.000000  ETH  ETHUSDC   
0     0.000010  USDC  0.000000  ETH  ETHUSDC   
0     0.000010  USDC  0.000000  ETH  ETHUSDC   
0     0.000010  USDC  0.000000  ETH  ETHUSDC   
0     0.000010  USDC  0.000000  ETH  ETHUSDC   
0     0.000010  USDC  0.000000  ETH  ETHUSDC   
0     0.000010  USDC  0.000000  ETH  ETHUSDC   
0     0.000010  USDC  0.000000  ETH  ETHUSDC   
0     0.000010  USDC  0.000000  ETH  ETHUSDC   
0     0.000010  USDC  0.000000  ETH  ETHUSDC   
0     0.000010  USDC  0.000000  ETH  ETHUSDC   
0    93.176617  USDC  0.033333  ETH  ETHUSDC   
0    93.176617  USDC  0.033333  ETH  ETHUSDC   
0    93.176617  USDC  0.033333  ETH  ETHUSDC   
0    93.176617  USDC  0.033333  ETH  ETHUSDC   
0    93.176617  USDC  0.033333  ETH  ETHUSDC   
0    93.176617  USDC  0.033333  ETH  ETHUSDC   
0    93.176617  USDC  0.033333  ETH  ETHUSDC   
0    93.176617  USDC  0.033333  ETH  ETHUSDC   
0    93.176617  USDC  0.033333  ETH  ETHUSDC   
0    93.176617  USDC  0.033333  ETH  ETHUSDC   
0    93.176617  USDC  0.033333  ETH  ETHUSDC   
0    93.176617  USDC  0.033333  ETH  ETHUSDC   
0    93.176617  USDC  0.033333  ETH  ETHUSDC   
0    93.176617  USDC  0.033333  ETH  ETHUSDC   
0    93.176617  USDC  0.033333  ETH  ETHUSDC   
0  1397.649409  USDC  0.500000  ETH  ETHUSDC   

                                             routeix  nroutes     

In [12]:
AlphaSim.amm_buys('ETH',eth_trade_amount, execute=False, threshold_orders=3)['trades']  # route_trade_by_source

uid  id subid                          note   aggr   exec limitfail  \
0  1.0   1     0                     route #15  False  False      None   
0  1.1   1     1                      route #1  False  False      None   
0  1.2   1     2                     route #11  False  False      None   
0    1   1     A  AMM sells 1383USDC buys 0ETH   True  False      None   

          amt1  tkn1  amt2 tkn2     pair      routeix  nroutes        price  \
0     0.000010  USDC  0.00  ETH  ETHUSDC           15        1  2765.121734   
0   691.280429  USDC  0.25  ETH  ETHUSDC            1        1  2765.121734   
0   691.280429  USDC  0.25  ETH  ETHUSDC           11        1  2765.121734   
0  1382.560867  USDC  0.50  ETH  ETHUSDC  [15, 1, 11]        3  2765.121734   

         p_unit  threshold_orders  
0  USDC per ETH                 3  
0  USDC per ETH                 3  
0  USDC per ETH                 3  
0  USDC per ETH                 3

Again, we expect that routing through fewer orders should give a slightly worse price for the trader.

Since here the amm is buying then the trader is selling 0.5 ETH, then the Alpha price of 1388.32 USDC means that the return on 0.5 ETH sold is 1388.32 USDC. By comparison, the standard route gives a return of 1392.97 USDC which is slightly better for the trader (33bp)

In [13]:
(AlphaSim.state()["trades"].query("uid=='1'").iloc[0]["amt1"],
 Sim.state()["trades"].query("uid=='1'").iloc[0]["amt1"])

(1382.560867, 1397.649409)

In [14]:
10000*(AlphaSim.state()["trades"].query("uid=='1'").iloc[0]["amt1"]
       /Sim.state()["trades"].query("uid=='1'").iloc[0]["amt1"]-1)

-107.95655836749351

## Route by Target

#### AMM sells USDC

First we test the standard simulator which should route through all 5 positions

In [15]:
Sim.amm_sells('USDC',usdc_trade_amount, execute=False)['trades']  # route_trade_by_target

uid  id subid                          note   aggr   exec limitfail  \
0   2.0   2     0                      route #3  False  False      None   
0   2.1   2     1                     route #13  False  False      None   
0   2.2   2     2                     route #23  False  False      None   
0   2.3   2     3                     route #33  False  False      None   
0   2.4   2     4                     route #43  False  False      None   
0   2.5   2     5                     route #53  False  False      None   
0   2.6   2     6                     route #63  False  False      None   
0   2.7   2     7                     route #73  False  False      None   
0   2.8   2     8                     route #83  False  False      None   
0   2.9   2     9                     route #93  False  False      None   
0  2.10   2    10                    route #103  False  False      None   
0  2.11   2    11                    route #113  False  False      None   
0  2.12   2    12                    route #123  False  False      None   
0  2.13   2    13                    route #133  False  False      None   
0  2.14   2    14                    route #143  False  False      None   
0  2.15   2    15                      route #1  False  False      None   
0  2.16   2    16                     route #11  False  False      None   
0  2.17   2    17                     route #21  False  False      None   
0  2.18   2    18                     route #31  False  False      None   
0  2.19   2    19                     route #41  False  False      None   
0  2.20   2    20                     route #51  False  False      None   
0  2.21   2    21                     route #61  False  False      None   
0  2.22   2    22                     route #71  False  False      None   
0  2.23   2    23                     route #81  False  False      None   
0  2.24   2    24                     route #91  False  False      None   
0  2.25   2    25                    route #101  False  False      None   
0  2.26   2    26                    route #111  False  False      None   
0  2.27   2    27                    route #121  False  False      None   
0  2.28   2    28                    route #131  False  False      None   
0  2.29   2    29                    route #141  False  False      None   
0     2   2     A  AMM sells 1000USDC buys 0ETH   True  False      None   

          amt1  tkn1      amt2 tkn2     pair  \
0     0.000010  USDC  0.000000  ETH  ETHUSDC   
0     0.000010  USDC  0.000000  ETH  ETHUSDC   
0     0.000010  USDC  0.000000  ETH  ETHUSDC   
0     0.000010  USDC  0.000000  ETH  ETHUSDC   
0     0.000010  USDC  0.000000  ETH  ETHUSDC   
0     0.000010  USDC  0.000000  ETH  ETHUSDC   
0     0.000010  USDC  0.000000  ETH  ETHUSDC   
0     0.000010  USDC  0.000000  ETH  ETHUSDC   
0     0.000010  USDC  0.000000  ETH  ETHUSDC   
0     0.000010  USDC  0.000000  ETH  ETHUSDC   
0     0.000010  USDC  0.000000  ETH  ETHUSDC   
0     0.000010  USDC  0.000000  ETH  ETHUSDC   
0     0.000010  USDC  0.000000  ETH  ETHUSDC   
0     0.000010  USDC  0.000000  ETH  ETHUSDC   
0     0.000010  USDC  0.000000  ETH  ETHUSDC   
0    66.666657  USDC  0.023838  ETH  ETHUSDC   
0    66.666657  USDC  0.023838  ETH  ETHUSDC   
0    66.666657  USDC  0.023838  ETH  ETHUSDC   
0    66.666657  USDC  0.023838  ETH  ETHUSDC   
0    66.666657  USDC  0.023838  ETH  ETHUSDC   
0    66.666657  USDC  0.023838  ETH  ETHUSDC   
0    66.666657  USDC  0.023838  ETH  ETHUSDC   
0    66.666657  USDC  0.023838  ETH  ETHUSDC   
0    66.666657  USDC  0.023838  ETH  ETHUSDC   
0    66.666657  USDC  0.023838  ETH  ETHUSDC   
0    66.666657  USDC  0.023838  ETH  ETHUSDC   
0    66.666657  USDC  0.023838  ETH  ETHUSDC   
0    66.666657  USDC  0.023838  ETH  ETHUSDC   
0    66.666657  USDC  0.023838  ETH  ETHUSDC   
0    66.666657  USDC  0.023838  ETH  ETHUSDC   
0  1000.000000  USDC  0.357572  ETH  ETHUSDC   

                                             routeix  nroutes     

Then we can set a threshold on the number of orders using the AlphaSim

In [16]:
AlphaSim.amm_sells('USDC',usdc_trade_amount, execute=False, threshold_orders = 3)['trades'] # route_trade_by_target

uid  id subid                          note   aggr   exec limitfail  \
0  2.0   2     0                      route #1  False  False      None   
0  2.1   2     1                     route #11  False  False      None   
0  2.2   2     2                     route #21  False  False      None   
0    2   2     A  AMM sells 1000USDC buys 0ETH   True  False      None   

          amt1  tkn1      amt2 tkn2     pair      routeix  nroutes  \
0   333.333333  USDC  0.119767  ETH  ETHUSDC            1        1   
0   333.333333  USDC  0.119767  ETH  ETHUSDC           11        1   
0   333.333333  USDC  0.119767  ETH  ETHUSDC           21        1   
0  1000.000000  USDC  0.359301  ETH  ETHUSDC  [1, 11, 21]        3   

         price        p_unit  threshold_orders  
0  2783.181806  USDC per ETH                 3  
0  2783.181806  USDC per ETH                 3  
0  2783.181806  USDC per ETH                 3  
0  2783.181806  USDC per ETH                 3

We would expect that routing through fewer orders should give a slightly worse price for the trader.

Since here the amm is selling then the trader is buying 1000 USDC, then the Alpha price of 0.359301 ETH means the trader is required to pay more ETH for the same 1000 USDC than the standard approach (0.358435 ETH).

This means that the higher ETH amount is a worse price as this is the amount that the trader will part with for a fixed target USDC amount (24bp)

In [17]:
(AlphaSim.state()["trades"].query("uid=='2'").iloc[0]["amt2"],
 Sim.state()["trades"].query("uid=='2'").iloc[0]["amt2"])

(0.359301, 0.357572)

In [18]:
10000*(AlphaSim.state()["trades"].query("uid=='2'").iloc[0]["amt2"]
       /Sim.state()["trades"].query("uid=='2'").iloc[0]["amt2"]-1)

48.35389795621481

#### AMM sells ETH

In [19]:
Sim.amm_sells('ETH',eth_trade_amount, execute=False)['trades']  # route_trade_by_target

uid  id subid                          note   aggr   exec limitfail  \
0   3.0   3     0                      route #6  False  False      None   
0   3.1   3     1                     route #16  False  False      None   
0   3.2   3     2                     route #26  False  False      None   
0   3.3   3     3                     route #36  False  False      None   
0   3.4   3     4                     route #46  False  False      None   
0   3.5   3     5                     route #56  False  False      None   
0   3.6   3     6                     route #66  False  False      None   
0   3.7   3     7                     route #76  False  False      None   
0   3.8   3     8                     route #86  False  False      None   
0   3.9   3     9                     route #96  False  False      None   
0  3.10   3    10                    route #106  False  False      None   
0  3.11   3    11                    route #116  False  False      None   
0  3.12   3    12                    route #126  False  False      None   
0  3.13   3    13                    route #136  False  False      None   
0  3.14   3    14                    route #146  False  False      None   
0  3.15   3    15                      route #0  False  False      None   
0  3.16   3    16                     route #10  False  False      None   
0  3.17   3    17                     route #20  False  False      None   
0  3.18   3    18                     route #30  False  False      None   
0  3.19   3    19                     route #40  False  False      None   
0  3.20   3    20                     route #50  False  False      None   
0  3.21   3    21                     route #60  False  False      None   
0  3.22   3    22                     route #70  False  False      None   
0  3.23   3    23                     route #80  False  False      None   
0  3.24   3    24                     route #90  False  False      None   
0  3.25   3    25                    route #100  False  False      None   
0  3.26   3    26                    route #110  False  False      None   
0  3.27   3    27                    route #120  False  False      None   
0  3.28   3    28                    route #130  False  False      None   
0  3.29   3    29                    route #140  False  False      None   
0     3   3     A  AMM sells 0ETH buys 1000USDC   True  False      None   

       amt1 tkn1         amt2  tkn2     pair  \
0  0.000010  ETH     0.019500  USDC  ETHUSDC   
0  0.000010  ETH     0.019500  USDC  ETHUSDC   
0  0.000010  ETH     0.019500  USDC  ETHUSDC   
0  0.000010  ETH     0.019500  USDC  ETHUSDC   
0  0.000010  ETH     0.019500  USDC  ETHUSDC   
0  0.000010  ETH     0.019500  USDC  ETHUSDC   
0  0.000010  ETH     0.019500  USDC  ETHUSDC   
0  0.000010  ETH     0.019500  USDC  ETHUSDC   
0  0.000010  ETH     0.019500  USDC  ETHUSDC   
0  0.000010  ETH     0.019500  USDC  ETHUSDC   
0  0.000010  ETH     0.019500  USDC  ETHUSDC   
0  0.000010  ETH     0.019500  USDC  ETHUSDC   
0  0.000010  ETH     0.019500  USDC  ETHUSDC   
0  0.000010  ETH     0.019500  USDC  ETHUSDC   
0  0.000010  ETH     0.019500  USDC  ETHUSDC   
0  0.033323  ETH    66.670121  USDC  ETHUSDC   
0  0.033323  ETH    66.670121  USDC  ETHUSDC   
0  0.033323  ETH    66.670121  USDC  ETHUSDC   
0  0.033323  ETH    66.670121  USDC  ETHUSDC   
0  0.033323  ETH    66.670121  USDC  ETHUSDC   
0  0.033323  ETH    66.670121  USDC  ETHUSDC   
0  0.033323  ETH    66.670121  USDC  ETHUSDC   
0  0.033323  ETH    66.670121  USDC  ETHUSDC   
0  0.033323  ETH    66.670121  USDC  ETHUSDC   
0  0.033323  ETH    66.670121  USDC  ETHUSDC   
0  0.033323  ETH    66.670121  USDC  ETHUSDC   
0  0.033323  ETH    66.670121  USDC  ETHUSDC   
0  0.033323  ETH    66.670121  USDC  ETHUSDC   
0  0.033323  ETH    66.670121  USDC  ETHUSDC   
0  0.033323  ETH    66.670121  USDC  ETHUSDC   
0  0.500000  ETH  1000.344322  USDC  ETHUSDC   

                                             routeix  nroutes     

In [20]:
AlphaSim.amm_sells('ETH',eth_trade_amount, execute=False, threshold_orders = 3)['trades'] # route_trade_by_target

uid  id subid                          note   aggr   exec limitfail  \
0  3.0   3     0                     route #16  False  False      None   
0  3.1   3     1                     route #26  False  False      None   
0  3.2   3     2                      route #0  False  False      None   
0    3   3     A  AMM sells 0ETH buys 1005USDC   True  False      None   

      amt1 tkn1         amt2  tkn2     pair      routeix  nroutes  \
0  0.00001  ETH     0.019500  USDC  ETHUSDC           16        1   
0  0.00001  ETH     0.019500  USDC  ETHUSDC           26        1   
0  0.49998  ETH  1005.266227  USDC  ETHUSDC            0        1   
0  0.50000  ETH  1005.305227  USDC  ETHUSDC  [16, 26, 0]        3   

         price        p_unit  threshold_orders  
0  2010.610453  USDC per ETH                 3  
0  2010.610453  USDC per ETH                 3  
0  2010.610453  USDC per ETH                 3  
0  2010.610453  USDC per ETH                 3

Again, we expect that routing through fewer orders should give a slightly worse price for the trader.

Since here the amm is selling then the trader is buying 0.5 ETH, then the Alpha price of 1001.762648 USDC means the trader is required to pay more USDC for the same 0.5 ETH than the standard approach (1001.056844 USDC).

This means that the higher USDC amount is a worse price as this is the amount that the trader will part with for a fixed target ETH amount (7bp)

In [21]:
(AlphaSim.state()["trades"].query("uid=='3'").iloc[0]["amt2"],
 Sim.state()["trades"].query("uid=='3'").iloc[0]["amt2"])

(1005.305227, 1000.344322)

In [22]:
10000*(AlphaSim.state()["trades"].query("uid=='3'").iloc[0]["amt2"]
       /Sim.state()["trades"].query("uid=='3'").iloc[0]["amt2"]-1)

49.591974392193535

### Decreasing the threshold number of orders

It follows that if we decrease the threshold then the price should get worse and vice-versa

In [23]:
AlphaSim.amm_sells('ETH',eth_trade_amount, execute=False, threshold_orders = 2)['trades'] # route_trade_by_target

uid  id subid                          note   aggr   exec limitfail  \
0  4.0   4     0                     route #16  False  False      None   
0  4.1   4     1                      route #0  False  False      None   
0    4   4     A  AMM sells 0ETH buys 1005USDC   True  False      None   

      amt1 tkn1        amt2  tkn2     pair  routeix  nroutes        price  \
0  0.00001  ETH     0.01950  USDC  ETHUSDC       16        1  2010.611879   
0  0.49999  ETH  1005.28644  USDC  ETHUSDC        0        1  2010.611879   
0  0.50000  ETH  1005.30594  USDC  ETHUSDC  [16, 0]        2  2010.611879   

         p_unit  threshold_orders  
0  USDC per ETH                 2  
0  USDC per ETH                 2  
0  USDC per ETH                 2

In [24]:
AlphaSim.amm_sells('ETH',eth_trade_amount, execute=False, threshold_orders = 4)['trades'] # route_trade_by_target

uid  id subid                          note   aggr   exec limitfail  \
0  5.0   5     0                     route #16  False  False      None   
0  5.1   5     1                     route #26  False  False      None   
0  5.2   5     2                     route #36  False  False      None   
0  5.3   5     3                      route #0  False  False      None   
0    5   5     A  AMM sells 0ETH buys 1005USDC   True  False      None   

      amt1 tkn1         amt2  tkn2     pair          routeix  nroutes  \
0  0.00001  ETH     0.019500  USDC  ETHUSDC               16        1   
0  0.00001  ETH     0.019500  USDC  ETHUSDC               26        1   
0  0.00001  ETH     0.019500  USDC  ETHUSDC               36        1   
0  0.49997  ETH  1005.246014  USDC  ETHUSDC                0        1   
0  0.50000  ETH  1005.304514  USDC  ETHUSDC  [16, 26, 36, 0]        4   

         price        p_unit  threshold_orders  
0  2010.609028  USDC per ETH                 4  
0  2010.609028  USDC per ETH                 4  
0  2010.609028  USDC per ETH                 4  
0  2010.609028  USDC per ETH                 4  
0  2010.609028  USDC per ETH                 4

And indeed reducing to 2 orders gives a worse price (i.e. costs more) and increasing to 4 orders gives a better price (i.e. costs less; 13bp)

In [25]:
(AlphaSim.state()["trades"].query("uid=='4'").iloc[0]["amt2"],
 AlphaSim.state()["trades"].query("uid=='5'").iloc[0]["amt2"])

(1005.30594, 1005.304514)

In [26]:
10000*(AlphaSim.state()["trades"].query("uid=='4'").iloc[0]["amt2"]
       / AlphaSim.state()["trades"].query("uid=='5'").iloc[0]["amt2"]-1)

0.014184756758073291

Generate 10,000 orders

In [35]:
for _ in range(2000):
    # Sim.add_strategy(*strat)
    AlphaSim.add_strategy(*strat)
    # Sim.add_strategy(*dusta)
    AlphaSim.add_strategy(*dusta)
    # Sim.add_strategy(*dustb)
    AlphaSim.add_strategy(*dustb)
    # Sim.add_strategy(*dustc)
    AlphaSim.add_strategy(*dustc)
    # Sim.add_strategy(*dustd)
    AlphaSim.add_strategy(*dustd)

#Sim.state()['orders']
AlphaSim.state()['orders']

id     pair   tkn       y_int           y y_unit  disabled  p_start  \
0          0  ETHUSDC   ETH    10.00000    10.00000    ETH     False   2000.0   
1          1  ETHUSDC  USDC  1000.00000  1000.00000   USDC     False   2800.0   
2          2  ETHUSDC   ETH     0.00001     0.00001    ETH     False   2050.0   
3          3  ETHUSDC  USDC     0.00001     0.00001   USDC     False   2850.0   
4          4  ETHUSDC   ETH     0.00001     0.00001    ETH     False   2550.0   
...      ...      ...   ...         ...         ...    ...       ...      ...   
11145  11145  ETHUSDC  USDC     0.00001     0.00001   USDC     False   2750.0   
11146  11146  ETHUSDC   ETH     0.00001     0.00001    ETH     False   1950.0   
11147  11147  ETHUSDC  USDC     0.00001     0.00001   USDC     False   2750.0   
11148  11148  ETHUSDC   ETH     0.00001     0.00001    ETH     False   2650.0   
11149  11149  ETHUSDC  USDC     0.00001     0.00001   USDC     False   2650.0   

        p_end  p_marg        p_unit    lid  
0      2500.0  2000.0  USDC per ETH      1  
1      2700.0  2800.0  USDC per ETH      0  
2      2050.0  2050.0  USDC per ETH      3  
3      2850.0  2850.0  USDC per ETH      2  
4      2550.0  2550.0  USDC per ETH      5  
...       ...     ...           ...    ...  
11145  2750.0  2750.0  USDC per ETH  11144  
11146  1950.0  1950.0  USDC per ETH  11147  
11147  2750.0  2750.0  USDC per ETH  11146  
11148  2650.0  2650.0  USDC per ETH  11149  
11149  2650.0  2650.0  USDC per ETH  11148  

[11150 rows x 12 columns]

In [39]:
AlphaSim.amm_sells('ETH',eth_trade_amount, execute=False, threshold_orders = 10)['trades'] # route_trade_by_target

uid  id subid                          note   aggr   exec limitfail  \
0  12.0  12     0                     route #16  False  False      None   
0  12.1  12     1                     route #26  False  False      None   
0  12.2  12     2                     route #36  False  False      None   
0  12.3  12     3                     route #46  False  False      None   
0  12.4  12     4                     route #56  False  False      None   
0  12.5  12     5                      route #0  False  False      None   
0    12  12     A  AMM sells 0ETH buys 1005USDC   True  False      None   

      amt1 tkn1         amt2  tkn2     pair                  routeix  nroutes  \
0  0.00001  ETH     0.019500  USDC  ETHUSDC                       16        1   
0  0.00001  ETH     0.019500  USDC  ETHUSDC                       26        1   
0  0.00001  ETH     0.019500  USDC  ETHUSDC                       36        1   
0  0.00001  ETH     0.019500  USDC  ETHUSDC                       46        1   
0  0.00001  ETH     0.019500  USDC  ETHUSDC                       56        1   
0  0.49995  ETH  1005.205588  USDC  ETHUSDC                        0        1   
0  0.50000  ETH  1005.303088  USDC  ETHUSDC  [16, 26, 36, 46, 56, 0]        6   

         price        p_unit  threshold_orders  
0  2010.606177  USDC per ETH                 6  
0  2010.606177  USDC per ETH                 6  
0  2010.606177  USDC per ETH                 6  
0  2010.606177  USDC per ETH                 6  
0  2010.606177  USDC per ETH                 6  
0  2010.606177  USDC per ETH                 6  
0  2010.606177  USDC per ETH                 6

In [40]:
AlphaSim.state()['trades'].query('aggr==True')

uid  id subid                          note  aggr   exec limitfail  \
0   0   0     A  AMM sells 0ETH buys 1000USDC  True  False      None   
0   1   1     A  AMM sells 1383USDC buys 0ETH  True  False      None   
0   2   2     A  AMM sells 1000USDC buys 0ETH  True  False      None   
0   3   3     A  AMM sells 0ETH buys 1005USDC  True  False      None   
0   4   4     A  AMM sells 0ETH buys 1005USDC  True  False      None   
0   5   5     A  AMM sells 0ETH buys 1005USDC  True  False      None   
0   6   6     A  AMM sells 0ETH buys 1005USDC  True  False      None   
0   7   7     A  AMM sells 0ETH buys 1005USDC  True  False      None   
0   8   8     A  AMM sells 0ETH buys 1005USDC  True  False      None   
0   9   9     A  AMM sells 0ETH buys 1005USDC  True  False      None   
0  10  10     A  AMM sells 0ETH buys 1005USDC  True  False      None   
0  11  11     A  AMM sells 0ETH buys 1005USDC  True  False      None   
0  12  12     A  AMM sells 0ETH buys 1005USDC  True  False      None   

          amt1  tkn1         amt2  tkn2     pair  \
0     0.497375   ETH  1000.000000  USDC  ETHUSDC   
0  1382.560867  USDC     0.500000   ETH  ETHUSDC   
0  1000.000000  USDC     0.359301   ETH  ETHUSDC   
0     0.500000   ETH  1005.305227  USDC  ETHUSDC   
0     0.500000   ETH  1005.305940  USDC  ETHUSDC   
0     0.500000   ETH  1005.304514  USDC  ETHUSDC   
0     0.500000   ETH  1005.304514  USDC  ETHUSDC   
0     0.500000   ETH  1005.300237  USDC  ETHUSDC   
0     0.500000   ETH  1005.236103  USDC  ETHUSDC   
0     0.500000   ETH  1005.243228  USDC  ETHUSDC   
0     0.500000   ETH  1005.271729  USDC  ETHUSDC   
0     0.500000   ETH  1005.300237  USDC  ETHUSDC   
0     0.500000   ETH  1005.303088  USDC  ETHUSDC   

                                             routeix  nroutes        price  \
0                                        [16, 26, 0]        3  2010.554434   
0                                        [15, 1, 11]        3  2765.121734   
0                                        [1, 11, 21]        3  2783.181806   
0                                        [16, 26, 0]        3  2010.610453   
0                                            [16, 0]        2  2010.611879   
0                                    [16, 26, 36, 0]        4  2010.609028   
0                                    [16, 26, 36, 0]        4  2010.609028   
0            [16, 26, 36, 46, 56, 66, 76, 86, 96, 0]       10  2010.600474   
0  [16, 26, 36, 46, 56, 66, 76, 86, 96, 106, 116,...      100  2010.472207   
0  [16, 26, 36, 46, 56, 66, 76, 86, 96, 106, 116,...       90  2010.486455   
0  [16, 26, 36, 46, 56, 66, 76, 86, 96, 106, 116,...       50  2010.543458   
0            [16, 26, 36, 46, 56, 66, 76, 86, 96, 0]       10  2010.600474   
0                            [16, 26, 36, 46, 56, 0]        6  2010.606177   

         p_unit  threshold_orders  
0  USDC per ETH                 3  
0  USDC per ETH                 3  
0  USDC per ETH                 3  
0  USDC per ETH                 3  
0  USDC per ETH                 2  
0  USDC per ETH                 4  
0  USDC per ETH                 4  
0  USDC per ETH                10  
0  USDC per ETH               100  
0  USDC per ETH                90  
0  USDC per ETH                50  
0  USDC per ETH                10  
0  USDC per ETH                 6